mono/mini/debugger-agent.c

```diff
@@ -4507,6 +4696,73 @@ set_breakpoint (MonoMethod *method, long il_offset, EventRequest *req, MonoError
 	return bp;
 }
 
+#ifdef IL2CPP_MONO_DEBUGGER
+
+static MonoBreakpoint* set_breakpoint_fast(Il2CppSequencePointC *sp, EventRequest *req, MonoError *error)
+{
+	MonoBreakpoint *bp;
+	GHashTableIter iter, iter2;
+	MonoDomain *domain;
+	MonoMethod *m;
+	MonoSeqPointInfo *seq_points;
+	GPtrArray *methods;
+	GPtrArray *method_domains;
+	GPtrArray *method_seq_points;
+	int i;
+
+	if (error)
+		mono_error_init(error);
+
+	// FIXME:
+	// - suspend/resume the vm to prevent code patching problems
+	// - multiple breakpoints on the same location
+	// - dynamic methods
+	// - races
+
+	bp = g_new0(MonoBreakpoint, 1);
+	bp->method = *sp->method;
+	bp->il_offset = sp->ilOffset;
+	bp->req = req;
+	bp->children = g_ptr_array_new();
+
+	DEBUG_PRINTF(1, "[dbg] Setting %sbreakpoint at %s:0x%x.\n", (req->event_kind == EVENT_KIND_STEP) ? "single step " : "", bp->method ? mono_method_full_name(bp->method, TRUE) : "<all>", (int)bp->il_offset);
+
+	methods = g_ptr_array_new();
+	method_domains = g_ptr_array_new();
+	method_seq_points = g_ptr_array_new();
+
+	mono_loader_lock();
+
+	BreakpointInstance* inst = g_new0(BreakpointInstance, 1);
+	inst->il_offset = bp->il_offset;// it.seq_point.il_offset;
+	inst->native_offset = 0;// it.seq_point.native_offset;
+	inst->domain = mono_domain_get();
+	inst->seq_point = sp;
+
+	sp->isActive = TRUE;
+
+	mono_loader_lock();
```

Comment:
We're taking this lock twice in a row without unlocking...I think thats undefined behavior isn't it

Reply:
I think that this is fine. We can take the loader lock twice in a row (it is a critical section on Windows, for example). We just need to make sure we unlock twice in a row, which this code does.